ApiDoctor.Validation/Http/HttpParser.cs

```diff
@@ -23,115 +23,118 @@
+                        var componentsLength = components.Length;
 
+                        if (componentsLength < 2)
+                            throw new HttpParserRequestException(
+                                "RequestString does not contain a proper HTTP request first line.");
                         if (components[0].StartsWith("HTTP/"))
                             throw new HttpParserRequestException("RequestString contains an HTTP response.");
 
                         request.Method = components[0];
-                        request.Url = components[1];
-
-                        if (components.Length >= 3)
-                        {
-                            request.HttpVersion = components[2];
-                        }
-                        else
-                        {
-                            request.HttpVersion = "HTTP/1.1";
-                        }
+                        request.Url = componentsLength > 3 ? line.Replace(request.Method, string.Empty) : components[1];
+                        request.HttpVersion = components[componentsLength - 1].StartsWith("HTTP/")
+                            ? components[componentsLength - 1]
+                            : "HTTP/1.1";
 
                         mode = ParserMode.Headers;
                         break;
 
                     case ParserMode.Headers:
-                        if (string.IsNullOrEmpty(line))
+                        if (string.IsNullOrEmpty(line) || line.StartsWith("{", true, CultureInfo.InvariantCulture))
                         {
+                            if (line.StartsWith("{", true, CultureInfo.InvariantCulture)) bodyBrace = line;
                             mode = ParserMode.Body;
                             continue;
                         }
 
                         // Parse each header
-                        int split = line.IndexOf(": ", StringComparison.Ordinal);
-                        if (split < 1) 
+                        var split = line.IndexOf(": ", StringComparison.Ordinal);
+                        if (split < 1)
                             throw new ArgumentException("requestString contains an invalid header definition");
 
                         var headerName = line.Substring(0, split);
                         var headerValue = line.Substring(split + 1);
                         request.Headers.Add(headerName, headerValue);
-
                         break;
 
                     case ParserMode.Body:
                         var restOfBody = reader.ReadToEnd() ?? string.Empty;
 
                         // normalize line endings to CRLF, which is required for headers, etc.
                         restOfBody = restOfBody.Replace("\r\n", "\n").Replace("\n", "\r\n");
-                        request.Body = string.Concat(line, "\r\n", restOfBody);
+                        request.Body = string.Concat(bodyBrace, line, "\r\n", restOfBody);
```

Comment:
why did you need to do the bodyBrace stuff?

Reply:
Corrected